---
title: "CNV-Calling-With-Exome-Depth"
author: "Alan Pittman"
date: "2/12/2020"
output: html_document
---

## R Markdown
## R Vession 3.6.1

#```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
#```

### Set Up

```{r}
library(ExomeDepth)
```

```{r}
setwd("/homedirs-porthos/sgul/shares/incc/porthos/Genetics_Centre_Bioinformatics/CNV_calling")
```


```{r}
data(exons.hg19)
print(head(exons.hg19))
```


##If happy with method go down to line 315 and start from there with your samples


## Test Data To Test Package !

```{r}
data(ExomeCount)
```

```{r}
ExomeCount.dafr <- as(ExomeCount, 'data.frame')

ExomeCount.dafr$chromosome <- gsub(as.character(ExomeCount.dafr$seqnames),
                                  pattern = 'chr',
                                  replacement = ' ') ##remove the annoying chr letters

```

```{r}
print(head(ExomeCount.dafr))
```

```{r}
test <- new('ExomeDepth',
          test = ExomeCount.dafr$Exome2,
          reference = ExomeCount.dafr$Exome3,
          formula = 'cbind(test, reference) ~ 1',
          subset.for.speed = seq(1, nrow(ExomeCount.dafr), 100))

```

## Build the most appropriate Reference set

```{r}
my.test <- ExomeCount$Exome4

my.ref.samples <- c('Exome1', 'Exome2', 'Exome3')

my.reference.set <- as.matrix(ExomeCount.dafr[, my.ref.samples])

my.choice <- select.reference.set (test.counts = my.test,
                                    reference.counts = my.reference.set,
                                    bin.length = (ExomeCount.dafr$end - ExomeCount.dafr$start)/1000,
                                    n.bins.reduced = 10000)

```

```{r}
my.matrix <- as.matrix( ExomeCount.dafr[, my.choice$reference.choice, drop = FALSE])

my.reference.selected <- apply(X = my.matrix,
                                MAR = 1,
                                FUN = sum)
```

## CNV calling

```{r}
all.exons <- new('ExomeDepth',
                  test = my.test,
                  reference = my.reference.selected,
                  formula = 'cbind(test, reference) ~ 1')
```


We can now call the CNV by running the underlying hidden Markov model:

```{r}
all.exons <- CallCNVs(x = all.exons,
                      transition.probability = 10^-4,
                      chromosome = ExomeCount.dafr$chromosome,
                      start = ExomeCount.dafr$start,
                      end = ExomeCount.dafr$end,
                      name = ExomeCount.dafr$names)
```

check output

```{r}
head(all.exons@CNV.calls)
```

No save it in an easily readable format


```{r}
data(Conrad.hg19)

head(Conrad.hg19.common.CNVs)
```


```{r}
all.exons <- AnnotateExtra(x = all.exons,
                          reference.annotation = Conrad.hg19.common.CNVs,
                          min.overlap = 0.5,
                          column.name = 'Conrad.hg19')
```

```{r}
print(head(all.exons@CNV.calls))
```


No save it in an easily readable format

```{r}
output.file <- 'Test_data_exome_calls.csv'
```

```{r}
write.csv(file = output.file,
                x = all.exons@CNV.calls,
                row.names = FALSE)
```


Complete!


###############################################################
###############################################################
###############################################################



### Now test some real data!


```{r}
analysisConfig <- read.csv('config.csv', 
							              header = TRUE, 
							              fill = TRUE)

```

```{r}
list_of_bam_files <- as.vector(analysisConfig$list_of_bam_files)

list_of_bam_files
```


#### Create count data from BAM files
#### Count for autosomal chromosomes


```{r}
my.counts <- getBamCounts(bed.frame = exons.hg19,
                          bam.files = list_of_bam_files,
                          include.chr = FALSE,)

```


```{r}
print(head(my.counts))
```


# Create dataframe
```{r}
ExomeCount.dafr <- as(my.counts[, colnames(my.counts)], 'data.frame')

print(head(ExomeCount.dafr))
```

# Create matrix of the bam counts
```{r}
ExomeCount.mat <- as.matrix(ExomeCount.dafr[, grep(names(ExomeCount.dafr), 
							pattern = '*.bam')])

print(head(ExomeCount.mat))
```							


```{r}
test2 <- new('ExomeDepth',
          test = ExomeCount.dafr$S28449_sorted_unique_recalibrated.bam,
          reference = ExomeCount.dafr$S32519_sorted_unique_recalibrated.bam,
          formula = 'cbind(test, reference) ~ 1',
          subset.for.speed = seq(1, nrow(ExomeCount.dafr), 100))

```

## Build the most appropriate Reference set

```{r}
my.test <- 'S28449_sorted_unique_recalibrated.bam'

my.test.data <- as.matrix(ExomeCount.dafr[, my.test])

my.ref.samples <- c('S32519_sorted_unique_recalibrated.bam', 'S29063_sorted_unique_recalibrated.bam')

my.reference.set <- as.matrix(ExomeCount.dafr[, my.ref.samples])


head(my.reference.set)

```

```{r}
my.choice <- select.reference.set (test.counts = my.test.data,
                                    reference.counts = my.reference.set,
                                    bin.length = (ExomeCount.dafr$end - ExomeCount.dafr$start)/1000,
                                    n.bins.reduced = 10000)

head(my.choice)

```


```{r}
my.matrix <- as.matrix( ExomeCount.dafr[, my.choice$reference.choice, drop = FALSE])

my.reference.set <- apply(X = my.matrix,
                                MAR = 1,
                                FUN = sum)
```

## CNV calling

```{r}
all.exons <- new('ExomeDepth',
                  test = ExomeCount.dafr$S28449_sorted_unique_recalibrated.bam,
                  reference = my.reference.set,
                  formula = 'cbind(test, reference) ~ 1')
```


We can now call the CNV by running the underlying hidden Markov model:

```{r}
all.exons <- CallCNVs(x = all.exons,
                      transition.probability = 10^-4,
                      chromosome = ExomeCount.dafr$chromosome,
                      start = ExomeCount.dafr$start,
                      end = ExomeCount.dafr$end,
                      name = ExomeCount.dafr$exon)
```

check output

```{r}
head(all.exons@CNV.calls)
```

Annotating with Conrad Common CNVs

```{r}
data(Conrad.hg19)

head(Conrad.hg19.common.CNVs)
```

```{r}
all.exons <- AnnotateExtra(x = all.exons,
                          reference.annotation = Conrad.hg19.common.CNVs,
                          min.overlap = 0.5,
                          column.name = 'Conrad.hg19')
```

```{r}
print(head(all.exons@CNV.calls))
```

Now save it in an easily readable format

```{r}
output.file <- 'Test1_data_exome_calls.csv'
```

```{r}
write.csv(file = output.file,
                x = all.exons@CNV.calls,
                row.names = FALSE)
```


done!


######################################################################################
######################################################################################
######################################################################################
######################################################################################


##Now for some real data - with Config file and Loop!

```{r}
analysisConfig <- read.csv('config.csv', 
							              header = TRUE, 
							              fill = TRUE)

```

```{r}
list_of_bam_files <- as.vector(analysisConfig$list_of_bam_files)

list_of_bam_files
```


#### Create count data from BAM files
#### Count for autosomal chromosomes


```{r}
my.counts <- getBamCounts(bed.frame = exons.hg19,
                          bam.files = list_of_bam_files,
                          include.chr = FALSE,)

```


```{r}
print(head(my.counts))
```


# Create dataframe
```{r}
ExomeCount.dafr <- as(my.counts[, colnames(my.counts)], 'data.frame')

print(head(ExomeCount.dafr))
```

# Create matrix of the bam counts
```{r}
ExomeCount.mat <- as.matrix(ExomeCount.dafr[, grep(names(ExomeCount.dafr), 
							pattern = '*.bam')])

print(head(ExomeCount.mat))
```							


```{r}
test <- new('ExomeDepth',
          test = ExomeCount.dafr$S28449_sorted_unique_recalibrated.bam,
          reference = ExomeCount.dafr$S32519_sorted_unique_recalibrated.bam,
          formula = 'cbind(test, reference) ~ 1',
          subset.for.speed = seq(1, nrow(ExomeCount.dafr), 100))

```

message('Now looping over all the samples innit')

```{r}
nsamples <- ncol(ExomeCount.mat)

print(head(nsamples))
```

```{r}
for (i in 1:nsamples) {

## Build the most appropriate Reference set


#my.test <- 'S28449_sorted_unique_recalibrated.bam'

my.test.data <- as.matrix(ExomeCount.mat[, i])

#my.ref.samples <- c('S32519_sorted_unique_recalibrated.bam', 'S29063_sorted_unique_recalibrated.bam')

my.reference.set <- as.matrix(ExomeCount.mat[, -i])

head(my.test.data)
head(my.reference.set)


my.choice <- select.reference.set(test.counts = my.test.data,
                                    reference.counts = my.reference.set,
                                    bin.length = (ExomeCount.dafr$end - ExomeCount.dafr$start)/1000,
                                    n.bins.reduced = 10000)



head(my.choice)

my.matrix <- as.matrix( ExomeCount.dafr[, my.choice$reference.choice, drop = FALSE])



my.reference.set <- apply(X = my.matrix,
                                MAR = 1,
                                FUN = sum)


## CNV calling


all.exons <- new('ExomeDepth',
                  test = ExomeCount.dafr$S28449_sorted_unique_recalibrated.bam,
                  reference = my.reference.set,
                  formula = 'cbind(test, reference) ~ 1')



#We can now call the CNV by running the underlying hidden Markov model:


all.exons <- CallCNVs(x = all.exons,
                      transition.probability = 10^-4,
                      chromosome = ExomeCount.dafr$space,
                      start = ExomeCount.dafr$start,
                      end = ExomeCount.dafr$end,
                      name = ExomeCount.dafr$names)




#check output


head(all.exons@CNV.calls)


#Annotating with Conrad Common CNVs


data(Conrad.hg19)

head(Conrad.hg19.common.CNVs)



all.exons <- AnnotateExtra(x = all.exons,
                          reference.annotation = Conrad.hg19.common.CNVs,
                          min.overlap = 0.5,
                          column.name = 'Conrad.hg19')



print(head(all.exons@CNV.calls))


#Now save it in an easily readable format


output.file <- paste('Exome_', 
					          i, 
					          '_', 	
					          list_of_bam_files[i], 
					          '.csv', sep = '')

write.csv(file = output.file, 
		      x = all.exons@CNV.calls, 
		      row.names = FALSE)

}

```

done!
